README.md
# Introduction
This repository contains code for the COVID-19 Response Dashboard developed by Mathematica with The Rockefeller Foundation .  

This dashboard tracks Covid-19 case counts, testing and vaccination response, and other contextual indicators in regions where The Rockefeller Foundation and its partners operate. Indicators are pulled or calculated from publicly available sources. For the United States, vaccine equity is a major focus of the dashboard. U.S. vaccination rates by race/ethnicity are calculated or imputed based on underlying data from the Centers for Disease Control and Prevention, the Kaiser Family Foundation, and the U.S. Census Bureau. Detailed source information is provided in the `data/df_dashboard_data_dictionary.xlsx` and `www/2022-02-11 data sources description.pdf` files.  

For more information, please contact Laura Blue at LBlue@mathematica-mpr.com and Anitha Sivasankaran at ASivasankaran@mathematica-mpr.com  

# Installation

## Software
To run the code in this repository you will need the following software installed:
* [Docker](https://docs.docker.com/get-docker/)

## Installation instructions
1. Run `git config core.hooksPath hooks/` to activate the scripts in the hooks directory. These hooks prevent the following:
* Forgetting to update renv.lock when package changes are made
* Forgetting to update the data file when the code is changed or updated.

2. *.csv files are tracked using Git Large File Storage. To install and use:
  
  * Download here: https://git-lfs.github.com/
  * Change `git config lfs.activitytimeout 3000`
  * To add other files to lfs, run 'git lfs track "*.fileExtension"'.
  * Make sure .gitattributes is tracked: `git add .gitattributes`

3. Docker

This project uses Docker, so you can run (or develop) the app in RStudio or deploy the app locally with the Dockerfile. Since this project using R 3.5, this is especially helpful with the upgrade to R 4.0 (otherwise, make sure to set Tools --> Global Options --> R Version to R 3.5)

To build the Docker image, git bash into the main project folder and run 
`docker build -t nameforimage .` 
where nameforimage is what you want to name this image 

Then to run a container from this image, run 
`docker run -d -e PASSWORD=myspecialpassword -p 8787:8787 -v local/path/to/folder:/home/rstudio nameforimage`
where local/path/to/folder is the path to the main project folder on your computer and nameforimage is the name of the image you created 

You can then go to `localhost:8787` to develop locally. Enter rstudio as your username and `myspecialpassword` (or whatever you set to be your password). 

If you want to view the app in shiny server, change the port mapping from 8787:8787 to 3838:3838 and then go to localhost:3838.

# Contents
* Dockerfile  
  + Dockerfile that contains the commands for a user to assemble a Docker image for local development and general deployment.
* data/
  + for storing intermediate and final datasets and dictionaries used by the Dashboard  
* src/
  + helper scripts for pulling and processing the data for the dashboard  
* create_df_dashboard.R  
  + main script for pulling and processing the data for the dashboard  
* ui.R  
  + Shiny App UI  
* server.R  
  + Shiny App Server
* www/
  + 2022-02-11 data sources description.pdf: PDF describing the data sources used in the Dashboard




HISTORY.md
# Project History

v6.6.0 (6/6/2022)
-----------------------
* refresh dashboard data

v6.5.0 (5/2/2022)
-----------------------
* refresh dashboard data

v6.4.0 (4/18/2022)
-----------------------
* refresh dashboard data

v6.3.0 (4/4/2022)
-----------------------
* refresh dashboard data

v6.2.0 (3/21/2022)
-----------------------
* refresh dashboard data

v6.1.1 (3/8/2022)
-----------------------
* refresh dashboard data

v6.0.1 (2/22/2022)
-----------------------
* refresh dashboard data

V6.0.0 (2/16/2022)
-----------------------
* Add vaccine acceptance panel on non-US country pages with indicators for: proportion who would accept a Covid-19 vaccine, vaccine eligibility, main reasons driving vaccine hesitancy, and structural barriers to vaccination among the unvaccinated
* Add vaccine acceptance panel on Region pages with indicator for proportion who would accept a Covid-19 vaccine
* Add figure for indicator for "Proportion vaccinated in all low- and middle-income countries globally (all ages, at least one dose)" on the Region page
* Remove case fatality ratio (CFR) indicator when there is insufficient data reported but displays indicator historically
* Add 12 month and 18 month "Show change over" options
* Fix: change Swaziland to Eswatini in the “geography-country-list.xlsx” 
* Fix: proportion BIPOC vaccinated among vaccinations w/ race/ethnicity reported to state agencies imputation 

v5.4.0 (2/8/2022)
-----------------------
* refresh dashboard data

v5.3.0 (1/24/2022)
-----------------------
* refresh dashboard data

v5.2.0 (1/10/2022)
-----------------------
* other as missing
* fixes to race/ethnicity national roll-up
* refresh dashboard data

v5.1.1 (12/27/2021)
-----------------------
* refresh dashboard data

v5.1.0 (12/17/2021)
-----------------------
* refresh dashboard data

v5.0.0 (11/29/2021)
-----------------------
* refresh dashboard data
* ux updates

v4.1.0 (11/15/2021)
-----------------------
* refresh dashboard data
* temp trim to CDC data to account for 5+ vaccinations (will change this in a future update)

v4.0.2 (11/9/2021)
-----------------------
* cleaned up data sources description

v4.0.1 (11/1/2021)
-----------------------
* user experience updates (basic visual updates)
* removed dashboard and non-US subnational pages
* refresh dashboard data

v3.19.0 (10/18/2021)
-----------------------
* top code case fatality ratio
* refresh dashboard data

v3.18.0 (10/4/2021)
-----------------------
* fix handling of missing OWID cumulative vaccinations by state when constructing total Total BIPOC population vaccinated, estimated
* refresh dashboard data

v3.17.0 (10/4/2021)
-----------------------
* refresh dashboard data

v3.16.0 (9/29/2021)
-----------------------
* added google analytics tracking

v3.15.0 (9/20/2021)
-----------------------
* refresh dashboard data

v3.14.0 (9/8/2021)
-----------------------
* hotfix to modify lookback to 60 days due to delayed KFF data on BIPOC vaccinations

v3.13.0 (9/7/2021)
-----------------------
* Refresh dashboard data
* Suppress PA Asian data
* Update renv.lock file

v3.12.0 (9/1/2021)
-----------------------
* Update disclaimer and about text

v3.11.1 (8/23/2021)
-----------------------
* refresh dashboard data
* suppress indian subnational data: data source link no longer works

v3.11.0 (8/9/2021)
-----------------------
* revise indicator definitions for evx targets
* refresh dashboard data

v3.10.0 (8/3/2021)
-----------------------
* update indicator definitions. add targets for evx indicators

v3.9.0 (7/25/2021)
-----------------------
* Refresh dashboard data
* simplify US page to focus on equitable vaccination
* improve dashboard speed
* add BIPOC proportion vaccinated to the core indicators

v3.8.0 (7/12/2021)
-----------------------
* Add disclaimer, new about text, RF logo

v3.7.1 (7/12/2021)
-----------------------
* remove Florida equity data before 6/13
* refresh dashboard data


v3.7.0 (6/30/2021)
-----------------------
* Include data fixes:
* don't assign zeros to us state-dates with missing data (affects CA)
* fix cdc county names
* filter out unused US geographies
* adjust y axis for equitable vaccination figures
* refresh dashboard data

v3.6.3 (6/28/2021)
-----------------------
* Hotfix: remove Florida data causing BIPOC disparity figure error

v3.6.2 (6/28/2021)
-----------------------
* Hotfix: re-do refresh of data through 06/27/2021

v3.6.1 (6/28/2021)
-----------------------
* Update: refresh data through 06/27/2021

v3.6.0 (6/15/2021)
-----------------------
* Set US country to be the landing page

v3.5.0 (6/15/2021)
-----------------------
* Update aesthetics
* Display latest policy indicators
* Add clickable pins to reach demonstration site pages

v3.4.1 (6/14/2021)
-----------------------
* Update: refresh data through 06/13/2021

v3.4.0 (6/8/2021)
-----------------------
* Update: refresh data through 06/07/2021
* Updated vaccine acceptance definition (includes vaccinated)
* Include BIPOC share of vaccinations figure

v3.3.2 (6/1/2021)
-----------------------
* Update: refresh data through 05/31/2021
* Update: mask anomalous Brazilian testing data on 5-24-2021

v3.3.1 (5/17/2021)
-----------------------
* Update: refresh data through 05/17/2021

v3.3.0 (5/14/2021)
-----------------------
* Update: include vaccine equity panel and demonstration sites, update documentation

ui.R
shinyUI(dashboardPage(

  title = "Rockefeller",
  dashboardHeader(
    title = uiOutput("dash_title"),
    tags$li(a(href = "https://www.rockefellerfoundation.org/", 
              img(src="RF white logo for dark backgrounds crop.png", 
                  title = "rockefellerfoundation.org", 
                  height = "50px"),
              style = "padding-top:10px; padding-bottom:10px;"), 
            class = "dropdown"),
    tags$li(a(href = "https://mathematica.org", 
              img(src="MathematicaLogo_White.png", 
                  title = "Mathematica.org", 
                  height = "50px"),
              style = "padding-top:10px; padding-bottom:10px;"), 
            class = "dropdown"),
    titleWidth = "550px"),
  
  dashboardSidebar(
    width = 350,
    useShinyalert(),
    collapsed = TRUE,
    br(), 
    sidebarMenu(id = "sidebar",
                menuItem(p("Dashboard", class = "menu-item", style = "font-family:'Montserrat'"), 
                         tabName = "dashboard", 
                         icon = icon('dashboard', class = "menu-item")),
                 div(conditionalPanel(condition = "input.sidebar == 'dashboard'", style = "padding-left: 15px;", 
                )),
                menuItem(p("Indicator definitions", class = "menu-item", style = "font-family:'Montserrat'"), 
                         tabName = "indicators", 
                         icon = icon('file', class = "menu-item")) 
    )
  ),

  dashboardBody(
    tags$head(includeCSS("www/styles.css")),
    tags$head(includeHTML(("google-analytics.html"))),
    tags$style(".fa-check {color:#4d4d4d}"),
    tags$style(".fa-minus {color:#4d4d4d}"),
    shinyjs::useShinyjs(),
    tabItems(
      tabItem(tabName = "dashboard", 
              
              fluidRow(
                p("About", class = "subheading", style = "margin:10px 0 2px 5px;"), 
                hr(),
                div(tags$p(about_text_1, class = "about-text"), style = "word-break:break-word; font-size: 16px;"),
                div(p("Options", class = "subheading", style = "margin:10px 0 2px 5px;")), 
                hr(),
                column(1),
                column(1,conditionalPanel(
                  condition = "input.level == 'Sub-national'",
                  uiOutput("country")
                )),
                column(2,
                 pickerInput(
                   inputId = "level",
                   p("Dashboard level", class = "input-label"),
                   selected = "Country",
                   choices = DASH_LEVELS,
                   options = list(`live-search` = TRUE))),
                column(2,uiOutput("geography"),style = "z-index:900;"),
                column(2,
                       sliderTextInput(
                         "date",
                         p("Date", class = "input-label"),
                         choices = DATE_RANGE,
                         selected = default_date,
                         grid = FALSE,
                         # hide_min_max = TRUE,
                         width = "95%"
                       )),
                 column(2,uiOutput("change_over")),
                 column(1),
                 column(1,conditionalPanel(
                    condition = "input.country == 'India' & input.level == 'Sub-national'",
                    uiOutput("district")
                  )),
                style = "background-color:#E0D4B5;"),
              fluidRow(
                # Line graphs
                conditionalPanel(
                  condition = "input.level != 'Demonstration site'",
                  column(6,
                         uiOutput("info_panel"),
                         h4(""),
                         uiOutput("date_panel"),
                         leafletOutput("map", height = 750),
                         style = "z-index:500;"),
                  column(1),
                  column(5,
                       conditionalPanel(
                        condition = "input.geography == 'United States' | (input.level == 'Sub-national' & input.country == 'United States')",
                        figTitle(title = "bipoc_pct_vacc_esttitle", info = bipoc_vacc_state_text),
                        plotlyOutput(outputId = "bipoc_pct_vacc_est_fig", height = "175px", width = "95%")
                       ),
                       figTitle(title = "vpeopletitle", info = vpeople_text),
                       plotlyOutput(outputId = "vpeople_fig", height = "190px", width = "95%"),
                       figTitle(title ="tprtitle", info = tpr_text), 
                       plotlyOutput(outputId = "tpr_fig", height = "175px", width = "95%"),
                       figTitle(title ="testtitle", info = tests_per_mil_text),
                       plotlyOutput(outputId = "tests_fig", height = "175px", width = "95%"),
                       figTitle(title ="casetitle", info = cases_per_mil_text),
                       plotlyOutput(outputId = "cases_fig", height = "175px", width = "95%"),
                       conditionalPanel(
                         condition = "input.level == 'Region' & input.geography != 'United States'",
                         figTitle(title ="lmictitle", info = lmic_text), 
                         plotlyOutput(outputId = "lmic_fig", height = "175px", width = "95%"))
                  )
                ),
                conditionalPanel(
                  condition = "input.level == 'Demonstration site'",
                  column(3,
                         uiOutput("demo_info_panel"),
                         h4(""),
                         uiOutput("demo_date_panel"),
                         leafletOutput("demo_map", height = 250)),
                  column(9,
                         figTitle(title = "demo_vpeopletitle", info = demo_vpeople_text, font_size = 24),
                         validate_modOutput(name = "prop_vacc_county", fig_height = 225, err_height = 40, width = "80%"),
                         div(style = "height: 15px; background-color:'transparent';"),
                         figTitle(title = "bipoc_vacc_state_title", info = bipoc_vacc_state_text, font_size = 24),
                         validate_modOutput(name = "demo_pevx", fig_height = 225, err_height = 40, width = "80%"),
                         div(style = "height: 15px; background-color:'transparent';"),
                         figTitle(title = "race_state_title", info = race_state_text, font_size = 24),
                         validate_modOutput(name = "demo_revx", fig_height = 225, err_height = 40, width = "80%"),
                         div(style = "height: 15px; background-color:'transparent';"),
                         figTitle(title = "vacc_accept_msa_title", info = vacc_accept_msa_text, font_size = 24),
                         validate_modOutput(name = "vacc_accept_msa", fig_height = 225, err_height = 40, width = "80%")
                       )
                )
              ),
              # Vaccine acceptance panel for countries except US
              conditionalPanel(
                condition = "input.level != 'Demonstration site' & input.level != 'Region' & input.geography != 'United States' & !(input.level == 'Sub-national' & input.country == 'United States')",
                fluidRow(
                  column(12,
                         sectionHeader("Vaccine acceptance"),
                         column(6,
                                figTitle("vaxacctitle", info = prop_accept_vacc_text),
                                plotlyOutput(outputId = "vax_acc_fig",
                                             height = "200px", width = "95%"),
                                br(),
                                figTitle("vaxeligtitle", info = vacc_elig_text, width = "225px"),
                                plotlyOutput(outputId = "vacc_elig_progress_bar",
                                             height = "130px", width = "95%")
                         ),
                         column(6,
                                fluidRow(
                                  column(1, icon("frown", "fa-3x")),
                                  column(11, figTitle("vaxhesttitle", info = vacc_hest_text))
                                ),
                                fluidRow(
                                  uiOutput("top_three_hest")
                                ),
                                br(),
                                br(),
                                fluidRow(
                                  column(1, icon("calendar-check", "fa-3x")),
                                  column(11, figTitle("strucbartitle", info = struc_barr_text))
                                ),
                                fluidRow(
                                  uiOutput("top_three_barrier")
                                ),
                                br()
                         )
                  )
                  
                )
              ),
              # Vaccine acceptance for regions, except US
              conditionalPanel(
                condition = "input.level == 'Region' & input.geography != 'United States'",
                fluidRow(
                  column(12,
                         sectionHeader("Vaccine acceptance"),
                         column(6,
                                figTitle("vaxacctitleregional", info = prop_accept_vacc_text),
                                plotlyOutput(outputId = "vax_acc_fig_regional",
                                             height = "200px", width = "95%")
                         )
                  )
                )
              ),
              
              # Trend graphs
              conditionalPanel(
                condition = "input.level != 'Demonstration site' & input.geography != 'United States' & !(input.level == 'Sub-national' & input.country == 'United States')",
                fluidRow(
                  column(6,
                    sectionHeader("COVID-19 burden of disease", info = burden_of_disease_text),
                    trend_header("recent_date", "display_change"),
                    uiOutput("dpm_row"),
                    uiOutput("cfr_row")
                  ),
                  # Policy indicators
                  column(6,
                         conditionalPanel(
                           sectionHeader("Policies enacted", info = policy_text, up = TRUE),
                           condition = "input.level != 'Region'",
                           uiOutput("policy_matrix")
                         ),
                         
                         # ,
                         # conditionalPanel(
                         #   condition = "input.level == 'Region'",
                         #   kpiUI("Stringency Index", id = "polstr")
                         # )
                  )
              )),
              
              conditionalPanel(
                condition = "input.geography == 'United States' | (input.country == 'United States' & input.level == 'Sub-national')",
                fluidRow(
                  div(class = "section-div-slim", h3("Vaccine Equity", class = "section-text")),
                  column(6, 
                         altFigTitle("Total BIPOC population vaccinated, at least one dose",
                                     info = tevx_text),
                         tags$div(
                           id = "tevx_figure_parent",
                           style = "height: 400px; width: 97.5%",
                           plotlyOutput("tevx_figure")),
                         tags$div(
                           id = "tevx_err_parent",
                           class = "shiny-output-error-validation",
                           style = "height: 80px;",
                           textOutput("tevx_err"))),
                  column(6, 
                         altFigTitle("Proportion BIPOC vaccinated (ages 5+, at least one dose)",
                                     info = bipoc_vacc_state_text),
                         tags$div(
                           id = "pevx_figure_parent",
                           style = "height: 400px; width: 97.5%",
                           plotlyOutput("pevx_figure")),
                         tags$div(
                           id = "pevx_err_parent",
                           class = "shiny-output-error-validation",
                           style = "height: 80px;",
                           textOutput("pevx_err"))),
                  style = "padding-bottom: 25px;"
                ),
                fluidRow(
                  column(6,
                         altFigTitle("BIPOC share of vaccinations to date, at least one dose",
                                     info = evx_text),
                         tags$div(
                           id = "disp_figure_parent",
                           style = "height: 400px; width: 97.5%",
                           plotlyOutput("disp_figure")),
                         tags$div(
                           id = "disp_err_parent",
                           class = "shiny-output-error-validation",
                           style = "height: 80px;",
                           textOutput("disp_err"))
                  ),
                  column(6,
                         altFigTitle("Proportion vaccinated by race/ethnicity (ages 5+, at least one dose), estimated",
                                     info = race_state_text),
                         tags$div(
                           id = "revx_figure_parent",
                           style = "height: 400px; width: 97.5%",
                           plotlyOutput("revx_figure")),
                         tags$div(
                           id = "revx_err_parent",
                           class = "shiny-output-error-validation",
                           style = "height: 80px;",
                           textOutput("revx_err"))),
                  style = "padding-bottom: 50px;"
                ),
                fluidRow(
                  column(6,
                         altFigTitle(title = textOutput("pvwkrfiguretitle", inline = T),
                                     info = prop_raceeth_text),
                         tags$div(
                           id = "pvwkr_figure_parent",
                           style = "height: 100px;",
                           div(style = "height: 30px;"),
                           kpiUI_notitle(id = "pvwkr_figure")),
                         tags$div(
                           id = "pvwkr_err_parent",
                           class = "shiny-output-error-validation",
                           style = "height: 80px;",
                           textOutput("pvwkr_err")),
                         div(style = "height: 100px; background-color:'transparent';")
                  ),
                  column(6,
                         conditionalPanel(
                           condition = "input.geography == 'United States'",
                         altFigTitle("Proportion of vaccinations with race/ethnicity information reported to state agencies",
                                     info = prop_raceeth_text),
                          tags$div(
                            id = "pvwkr_figure_parent_kff",
                            style = "height: 100px;",
                            div(style = "height: 30px;"),
                            kpiUI_notitle(id = "pvwkr_figure_kff")),
                          tags$div(
                            id = "pvwkr_err_parent_kff",
                            class = "shiny-output-error-validation",
                            style = "height: 80px;",
                            textOutput("pvwkr_err_kff"))),
                  div(style = "height: 100px; background-color:'transparent';")
                  ),
                )
              ),
              conditionalPanel(
                condition = "input.level != 'Demonstration site' & input.geography != 'United States' & !(input.level == 'Sub-national' & input.country == 'United States')",
                fluidRow(
                  # Annual indicators
                  div(class = "section-div-slim", htmlOutput("kpi_header")),
                  column(6, style = "padding-left:0",
                         conditionalPanel(# hide cvi for subnational pages
                           condition = "input.level != 'Region' & !(input.geography == 'India' & input.level=='Country')",
                           kpiUI("cvititle", id = "cvi"),
                           div(style = "height: 50px;")
                         ),
                         kpiUI("ehsititle", id = "ehsi"),
                         conditionalPanel(# hide dtp3 for subnational pages
                           condition = "input.level != 'Sub-national'",
                           div(style = "height: 50px;"),
                           kpiUI("dtp3title", id = "dtp3")
                         )
                  ),
                  conditionalPanel(
                    condition = "input.geography == 'United States' | (input.country == 'United States' & input.level == 'Sub-national')",
                    column(6, 
                           uiOutput("inptitle"),
                           plotlyOutput("inp_figure", height = "220px"),
                           uiOutput("icutitle"),
                           plotlyOutput("icu_figure", height = "220px")
                           )
                  )
                )
           ),
           
           fluidRow(
             div(style = "height: 15px; background-color:'transparent';"),
             h2("MORE INFORMATION"),
             
             div(HTML(
               paste("<p class = about-text>",
                     about_text_3a,
                     "<a style='color:#6666FF; text-decoration:underline;' href=",
                     mailer,
                     ">",
                     about_text_3b,
                     " </a>", 
                     about_text_3c,
                     version_text,
                     "</p>")), 
               style = "word-break:break-word; font-size: 16px;"),
             
             column(3, 
                      actionButton("switchtab","Indicator definitions",
                                   icon = icon('file'),
                                   class = "shiny-download-link",
                                   style = "background-color:#E0D4B5; border-color: #E0D4B5")),
             column(4,tags$li(class = "menu-item", 
                     downloadLink(
                       "downloadPDF", 
                       tagList(icon('database', class = 'menu-item'), 
                               span(p("Download data source details", 
                                      class = "shiny-download-link", 
                                      style = "font-family:'Montserrat'"))) 
                     ))),
             style = "background-color:#E0D4B5;"
           ),
           
           fluidRow(
             div(style = "height: 15px; background-color:'transparent';"),
             h2("DISCLAIMER AND LIMITATION OF LIABILITY"),
             h4(disclaimer, style = "margin-left: 5px;"),
             style = "background-color:#E0D4B5;"
             )
           ),
          tabItem(tabName = "indicators", 
                  div(style = "background-color:white;padding:10px;border:1px solid #000000;", 
                    h2("Core indicators", style = "font-size:20px; text-align:center;"), 
                    tagList(core_indicators)
                    ), 
                  div(
                    h2("Supporting indicators", style = "font-size:20px; text-align:center;"),
                    tagList(sup_indicators))
                  )
  
  ) # tab items close
) # dashboard body close

) # dashboard page close
) # shinyui close


